cmd/kubelet/app/server.go

```diff
@@ -283,6 +165,141 @@ HTTP server: The kubelet can also listen for HTTP and respond to a simple API
+		if err := utilfeature.DefaultMutableFeatureGate.SetFromMap(kubeletConfig.FeatureGates); err != nil {
+			klog.Fatal(err)
+		}
+	}
+
+	// We always validate the local configuration (command line + config file).
+	// This is the default "last-known-good" config for dynamic config, and must always remain valid.
+	if err := kubeletconfigvalidation.ValidateKubeletConfiguration(kubeletConfig); err != nil {
+		klog.Fatal(err)
+	}
+
+	// use dynamic kubelet config, if enabled
+	var kubeletConfigController *dynamickubeletconfig.Controller
+	if dynamicConfigDir := kubeletFlags.DynamicConfigDir.Value(); len(dynamicConfigDir) > 0 {
+		var dynamicKubeletConfig *kubeletconfiginternal.KubeletConfiguration
+		var err error
+		dynamicKubeletConfig, kubeletConfigController, err = BootstrapKubeletConfigController(dynamicConfigDir,
+			func(kc *kubeletconfiginternal.KubeletConfiguration) error {
+				// Here, we enforce flag precedence inside the controller, prior to the controller's validation sequence,
+				// so that we get a complete validation at the same point where we can decide to reject dynamic config.
+				// This fixes the flag-precedence component of issue #63305.
+				// See issue #56171 for general details on flag precedence.
+				return kubeletConfigFlagPrecedence(kc, args)
+			})
+		if err != nil {
+			klog.Fatal(err)
+		}
+		// If we should just use our existing, local config, the controller will return a nil config
+		if dynamicKubeletConfig != nil {
+			kubeletConfig = dynamicKubeletConfig
+			// Note: flag precedence was already enforced in the controller, prior to validation,
+			// by our above transform function. Now we simply update feature gates from the new config.
+			if err := utilfeature.DefaultMutableFeatureGate.SetFromMap(kubeletConfig.FeatureGates); err != nil {
+				klog.Fatal(err)
+			}
+		}
+	}
+
+	// construct a KubeletServer from kubeletFlags and kubeletConfig
+	kubeletServer := &options.KubeletServer{
+		KubeletFlags:         *kubeletFlags,
+		KubeletConfiguration: *kubeletConfig,
+	}
+
+	// use kubeletServer to construct the default KubeletDeps
+	kubeletDeps, err := UnsecuredDependencies(kubeletServer)
+	if err != nil {
+		klog.Fatal(err)
+	}
+
+	// add the kubelet config controller to kubeletDeps
+	kubeletDeps.KubeletConfigController = kubeletConfigController
+
+	// start the experimental docker shim, if enabled
+	if kubeletServer.KubeletFlags.ExperimentalDockershim {
+		if err := RunDockershim(&kubeletServer.KubeletFlags, kubeletConfig, stopCh); err != nil {
+			klog.Fatal(err)
+		}
+		return
+	}
+
+	// run the kubelet
+	klog.V(5).Infof("KubeletConfiguration: %#v", kubeletServer.KubeletConfiguration)
+	if err := Run(kubeletServer, kubeletDeps, stopCh); err != nil {
+		klog.Fatal(err)
+	}
+}
+
+//handleCleanFlagSet handle cleanFlagSet, initial, check, etc
+//if cleanFlag is help return true, else return false
+func handleCleanFlagSet(cmd *cobra.Command, args []string, cleanFlagSet *pflag.FlagSet) bool {
```

Comment:
Functions should be as short as possible, and a function does one thing. The deal of CleanFlagSet can be completely independent, I think. Thanks for your review.

Reply:
That is a good ideal in general, but I don't agree with this specific application of that rule.

Though the operations are coupled to `cleanFlagSet`'s state, putting them all in `handleCleanFlagSet` obscures the fact that parsing the flags vs. checking for invalid commands vs. handling help text are relatively distinct operations - the function isn't really doing one thing. And other operations related to flags like `verflag.PrintAndExitIfRequested()` and `utilflag.PrintFlags(cleanFlagSet)` were left out of `handleCleanFlagSet`, so the boundary of what it should contain isn't really clear either.

On one hand, we could make helpers for each of these steps, but on the other hand, they would only ever be called in one place, in the same order as the existing code, which already has comments to signpost each step. So I don't think it's likely to significantly improve readability.

I definitely share your desire to abstract away a lot of this boilerplate logic and make it easier to write a bootstrap sequence, and I'm not trying to shut you down - you obviously have good instincts to want to clean this up :). 

We're putting together some plans to standardize things like this, possibly by writing a wrapper around pflag and cobra, so that individual components don't have to deal with things like this directly. You might want to check out these discussions:
- https://github.com/kubernetes/community/pull/2977
- [KEP 0032](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md)
- https://github.com/kubernetes/community/pull/3008

Reply:
Thanks for your patient explanation. Since there are some plans, I will close this pr.